lib/pages/goal_page.dart

```diff
@@ -3,102 +3,529 @@ import 'package:linear/model/goal.dart';
+import 'package:linear/pages/profile_page/profile_page.dart';
+import 'package:linear/model/goal.dart';
+ 
+// ignore: must_be_immutable
 class GoalPage extends StatefulWidget {
-  const GoalPage({super.key});
-
-  @override
-  State<GoalPage> createState() => GoalPageState();
+ const GoalPage({super.key});
+ 
+ 
+ @override
+ State<GoalPage> createState() => GoalPageState();
 }
-
+ 
 class GoalPageState extends State<GoalPage> {
-  List<dynamic> _goals = [];
+ 
+    List<dynamic> _allGoals= [];
+    List<dynamic> _goals = [];
+    String sortingBy = "Newest";
+    String filteringBy= "Open";
+    bool isSorting = true;
 
-  bool _isLoading = true;
-  bool _isErrorFetchingGoals = false;
+    TextEditingController searchInput = TextEditingController();
 
-  @override
-  void initState() {
-    super.initState();
-    fetchGoals();
-  }
+  List<dynamic> _searchResults = [];
+  bool _hasPerformedSearch = false;
+   bool _isSearching = false;
+ 
+ bool _isLoading = true;
+ bool _isErrorFetchingGoals = false;
+ 
+ @override
+ void initState() {
+   super.initState();
+   fetchGoals();
+ }
+ 
+ fetchGoals() {
+   final Future<Map<String, dynamic>> successfulMessage = api.getGoalsForGoalPage(context);
+   successfulMessage.then((response) {
+     if (response['status'] == true) {
+       setState(() {
+         _allGoals = response['goals'];
+         _isLoading = true;
+         _isErrorFetchingGoals = false;
+       });
+       setState(() {
+         _goals = _allGoals.toList();
+       });
+     getOpen(_goals);
+     isSorting = true;
+     sortingBy = "Newest";
+     filteringBy= "Open";
+     setState(() {
+         _isLoading = false;
+         _isErrorFetchingGoals = false;
+       });
+ 
+     } else {
+       setState(() {
+          _allGoals = [];
+         _isLoading = false;
+         _isErrorFetchingGoals = true;
+       });
+       showDialog(
+            context: context,
+            builder: (context) {
+              return AlertDialog(
+                title: const Text("Error!"),
+                content: const Text(
+                    "An error occured while attempting to get the goals"),
+                actions: [
+                  TextButton(
+                      onPressed: () {
+                        Navigator.pop(context);
+                      },
+                      child: const Text("Ok"))
+                ],
+              );
+            });
+     }
+   });
+ }
 
-  fetchGoals() {
-    final Future<Map<String, dynamic>> successfulMessage = api.getGoalsForGoalPage(context);
-    successfulMessage.then((response) {
-      if (response['status'] == true) {
-        setState(() {
-          _goals = response['goals'];
-          _isLoading = false;
-          _isErrorFetchingGoals = false;
-        });
-      } else {
-        setState(() {
-          _goals = [];
-          _isLoading = false;
-          _isErrorFetchingGoals = true;
-        });
-      }
-    });
-  }
+ getOpen(List getGoals){
```

Comment:
all of these filter methods (lines 91 to 165) seem to be doing mostly the same thing; instead do this:

1. I see we have the `filteringBy` variable, change that to an enum.
2. all of these functions can be replaced by a single `applyFilter()` function, where we'll apply the current `filteringBy ` filter

In addition to simplicity, this will be useful to be able to sort/filter search results